Use native file open dialogs

saivert opened the pull request:
Initial work to use native file open dialogs in GTK3 UI.

saivert committed:
Use native file open dialogs

kuba160 commented on plugins/gtkui/actionhandlers.c:
```diff
@@ -130,25 +130,75 @@ set_file_filter (GtkWidget *dlg, const char *name) {
 gboolean
 action_open_files_handler_cb (void *userdata) {
-    GtkWidget *dlg = gtk_file_chooser_dialog_new (_("Open file(s)..."), GTK_WINDOW (mainwin), GTK_FILE_CHOOSER_ACTION_OPEN, GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL, GTK_STOCK_OPEN, GTK_RESPONSE_OK, NULL);
+#if defined _WIN32 && GTK_CHECK_VERSION(3,20,0)
```
I'm using `__MINGW32__`  constant to check for windows compiler, it would be nice to keep it same over all files.

kuba160 commented on plugins/gtkui/actionhandlers.c:
```diff
@@ -130,25 +130,75 @@ set_file_filter (GtkWidget *dlg, const char *name) {
+#endif
 
+#ifdef _WIN32
+    set_file_filter_win32 (dlg, NULL);
```
as I understand gtk native filechooser requirest different filtering, or?
I'd suggest to merge this `#ifdef` block with the one over it, so that GTK2 version still uses normal set_file_filter

kuba160 commented on plugins/gtkui/actionhandlers.c:
```diff
@@ -171,24 +221,37 @@ action_open_files_handler (struct DB_plugin_action_s *action, int ctx) {
+    GtkWidget *dlg = gtk_file_chooser_dialog_new (_("Add file(s) to playlist..."), GTK_WINDOW (mainwin), GTK_FILE_CHOOSER_ACTION_OPEN, GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL, GTK_STOCK_OPEN, GTK_RESPONSE_ACCEPT, NULL);
+#endif
 
+#ifdef _WIN32
```
same as upper

kuba160 commented:
Binaries in case you need it: [ddb_gui.zip](https://github.com/kuba160/deadbeef/files/1732263/ddb_gui.zip)

As marked in changes requested, the set_file_filter_win32 for GTK2 plugin breaks filtering for it.

Now for GTK3: 
- Add files option now turns windows native file chooser, filtering works, only thing I can complain is that multiple tracks are being added in opposite order.

- Add folders option is still broken in GTK, for me this is a huge drawback for GTK3 plugin, also fixing it would be great. I don't really understand source of the problem of it. New version still produces crash, but when I wanted to debug it, I found out that it works when run deadbeef through gdb. So that's a big question mark for me. If you have any ideas why it's happening please tell.

Another thing, I don't have much GTK knowledge, but this happens after dialog close:
```(deadbeef.exe:8608): Gtk-CRITICAL **: gtk_widget_destroy: assertion 'GTK_IS_WIDGET (widget)' failed```

kuba160 commented:
I just found that crash doesn't happen when run it through my terminal (cmder), but happens when run through mingw default terminal.
Backtrace doesn't say much, I'll try to compile gtk with debug symbols.
```
Thread 1 received signal SIGTRAP, Trace/breakpoint trap.
0x00000000685f779f in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
(gdb) bt
#0  0x00000000685f779f in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
#1  0x00000000685f8314 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
#2  0x00000000685f869f in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
#3  0x00000000685f87c8 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
#4  0x000000006d647d08 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgio-2.0-0.dll
#5  0x0000000063a4b368 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#6  0x0000000063a4cce4 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#7  0x0000000063a4cf92 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#8  0x000000006654133c in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgtk-3-0.dll
#9  0x0000000066544233 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgtk-3-0.dll
#10 0x000000006653c946 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgtk-3-0.dll
#11 0x0000000063a46293 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#12 0x0000000063a57ee7 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#13 0x0000000063a5fd48 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#14 0x0000000063a607c1 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#15 0x000000006654ad52 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgtk-3-0.dll
#16 0x000000006654b085 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgtk-3-0.dll
#17 0x0000000063a4d1c4 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#18 0x0000000063a4dedb in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#19 0x0000000063a4d1c4 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#20 0x0000000063a4dedb in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#21 0x0000000063a4b3c4 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#22 0x0000000063a4cce4 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#23 0x0000000063a4cf92 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgobject-2.0-0.dll
#24 0x000000006653fc1a in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgtk-3-0.dll
#25 0x0000000002ad2794 in action_add_folders_handler_cb (
    user_data=<optimized out>) at actionhandlers.c:276
#26 0x0000000070ec175d in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgdk-3-0.dll
#27 0x00000000685f21e0 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
#28 0x00000000685f2538 in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
#29 0x00000000685f277d in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libglib-2.0-0.dll
#30 0x000000006d5e988d in ?? ()
   from C:\Users\Kuba\Desktop\deadbeef-x86_64\libgio-2.0-0.dll
#31 0x0000000002a8503c in gtkui_thread (ctx=ctx@entry=0x0) at gtkui.c:1400
#32 0x0000000002a850be in gtkui_start () at gtkui.c:1529
#33 0x000000000043bfd3 in main (argc=<optimized out>, argv=<optimized out>)
    at main.c:1271
```

saivert commented:
I have fixed the problem with reverse list by reversing the GSList returned from gtk_file_chooser_get_filenames. I guess this is a GTK bug. It is certainly not something I remember from using the Win32 API directly.

I forgot some extra IFDEFs for the other occurences of set_file_filter_win32 in the code. Fixed.

I get no crashes like you do. We need to compare build environments. I get larger .dll files than you.

Do you want me to put the remaining fixes in a single commit or  do you prefer if I amend this one (and force pusing to my repo)?

kuba160 commented:
I sent stripped build, that's why (don't know why I sent stripped, debug symbols would be more useful).
Just push it so that changes appear here, that's only thing I need.

saivert committed:
Requested fixes for native file open dialog

saivert committed:
Merge ifdefs
